diff --git a/xpra/x11/prop_conv.py b/xpra/x11/prop_conv.py
index 5607072a30..69d0b7bf37 100644
--- a/xpra/x11/prop_conv.py
+++ b/xpra/x11/prop_conv.py
@@ -58,7 +58,7 @@ class NetWMStrut:
         if len(data)==16:
             self.left, self.right, self.top, self.bottom = struct.unpack(b"@LLLL", data)
         else:
-            data = _force_length("_NET_WM_STRUT or _NET_WM_STRUT_PARTIAL", data, 4 * 12)
+            data = _force_length("_NET_WM_STRUT or _NET_WM_STRUT_PARTIAL", data, 8 * 12)
             (
                 self.left, self.right, self.top, self.bottom,
                 self.left_start_y, self.left_end_y,
